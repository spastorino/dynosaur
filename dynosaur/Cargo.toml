[package]
name = "dynosaur"
version = "0.2.0"
authors = [
    "Santiago Pastorino <spastorino@gmail.com>",
    "Tyler Mandry <tmandry@google.com>",
    "Niko Matsakis <rust@nikomatsakis.com>",
]
categories = ["asynchronous", "no-std"]
keywords = ["async", "trait", "impl"]
description = "Dynamic dispatch for return position impl traits and async in Rust"
documentation = "https://docs.rs/dynosaur"
license = "MIT OR Apache-2.0"
repository = "https://github.com/spastorino/dynosaur"
edition = "2021"
rust-version = "1.75"

[dependencies]
dynosaur_derive = { version = "0.2", path = "../dynosaur_derive" }

[dev-dependencies]
tokio = { workspace = true }
trait-variant = { workspace = true }
ui_test = "0.28"

[[example]]
name = "next"

[[example]]
name = "trait_variant"

[[test]]
name = "ui"
harness = false
