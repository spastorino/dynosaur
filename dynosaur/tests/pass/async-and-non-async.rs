use dynosaur::dynosaur;

#[dynosaur(DynJobQueue = dyn(box))]
pub trait JobQueue {
    fn len(&self) -> usize;
    async fn dispatch(&self);
}

fn main() {}
