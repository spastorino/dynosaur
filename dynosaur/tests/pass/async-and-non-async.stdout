#![feature(prelude_import)]
#[prelude_import]
use std::prelude::rust_2021::*;
#[macro_use]
extern crate std;
use dynosaur::dynosaur;

pub trait JobQueue {
    fn len(&self)
    -> usize;
    async fn dispatch(&self);
}
mod __dynosaur_macro_dynjobqueue {
    use super::*;
    pub trait ErasedJobQueue {
        fn len(&self)
        -> usize;
        fn dispatch<'life0, 'dynosaur>(&'life0 self)
        ->
            ::core::pin::Pin<Box<dyn ::core::future::Future<Output = ()> +
            'dynosaur>>
        where
        'life0: 'dynosaur,
        Self: 'dynosaur;
    }
    impl<DYNOSAUR: JobQueue> ErasedJobQueue for DYNOSAUR {
        fn len(&self) -> usize { <Self as JobQueue>::len(self) }
        fn dispatch<'life0, 'dynosaur>(&'life0 self)
            ->
                ::core::pin::Pin<Box<dyn ::core::future::Future<Output = ()> +
                'dynosaur>> where 'life0: 'dynosaur, Self: 'dynosaur {
            Box::pin(<Self as JobQueue>::dispatch(self))
        }
    }
    #[repr(transparent)]
    pub struct DynJobQueue<'dynosaur_struct> {
        ptr: dyn ErasedJobQueue + 'dynosaur_struct,
    }
    impl<'dynosaur_struct> JobQueue for DynJobQueue<'dynosaur_struct> {
        fn len(&self) -> usize { self.ptr.len() }
        fn dispatch(&self) -> impl ::core::future::Future<Output = ()> {
            let ret:
                    ::core::pin::Pin<Box<dyn ::core::future::Future<Output =
                    ()> + '_>> = self.ptr.dispatch();
            let ret:
                    ::core::pin::Pin<Box<dyn ::core::future::Future<Output =
                    ()> + 'static>> = unsafe { ::core::mem::transmute(ret) };
            ret
        }
    }
    impl<'dynosaur_struct> DynJobQueue<'dynosaur_struct> {
        pub fn new_box(value: impl JobQueue + 'dynosaur_struct)
            -> Box<DynJobQueue<'dynosaur_struct>> {
            let value = Box::new(value);
            let value: Box<dyn ErasedJobQueue + 'dynosaur_struct> = value;
            unsafe { ::core::mem::transmute(value) }
        }
        pub fn new_arc(value: impl JobQueue + 'dynosaur_struct)
            -> std::sync::Arc<DynJobQueue<'dynosaur_struct>> {
            let value = std::sync::Arc::new(value);
            let value: std::sync::Arc<dyn ErasedJobQueue + 'dynosaur_struct> =
                value;
            unsafe { ::core::mem::transmute(value) }
        }
        pub fn new_rc(value: impl JobQueue + 'dynosaur_struct)
            -> std::rc::Rc<DynJobQueue<'dynosaur_struct>> {
            let value = std::rc::Rc::new(value);
            let value: std::rc::Rc<dyn ErasedJobQueue + 'dynosaur_struct> =
                value;
            unsafe { ::core::mem::transmute(value) }
        }
        pub fn from_ref(value: &(impl JobQueue + 'dynosaur_struct))
            -> &DynJobQueue<'dynosaur_struct> {
            let value: &(dyn ErasedJobQueue + 'dynosaur_struct) = &*value;
            unsafe { ::core::mem::transmute(value) }
        }
        pub fn from_mut(value: &mut (impl JobQueue + 'dynosaur_struct))
            -> &mut DynJobQueue<'dynosaur_struct> {
            let value: &mut (dyn ErasedJobQueue + 'dynosaur_struct) =
                &mut *value;
            unsafe { ::core::mem::transmute(value) }
        }
    }
    impl<DYNOSAUR: JobQueue> JobQueue for Box<DYNOSAUR> where DYNOSAUR: ?Sized
        {
        fn len(&self) -> usize { DYNOSAUR::len(self) }
        fn dispatch(&self) -> impl ::core::future::Future<Output = ()> {
            DYNOSAUR::dispatch(self)
        }
    }
}
use __dynosaur_macro_dynjobqueue::DynJobQueue;

fn main() {}
